vault/seal_autoseal.go

```diff
@@ -499,3 +513,84 @@ func (d *autoSeal) migrateRecoveryConfig(ctx context.Context) error {
 
 	return nil
 }
+
+// StartHealthCheck starts a goroutine that tests the health of the auto-unseal backend once every 10 minutes.
+// If unhealthy, logs a warning on the condition and begins testing every one minute until healthy again.
+func (d *autoSeal) StartHealthCheck() {
+	d.StopHealthCheck()
+	d.hcLock.Lock()
+	defer d.hcLock.Unlock()
+
+	healthCheck := time.NewTicker(sealHealthTestIntervalNominal)
+	d.healthCheckStop = make(chan struct{})
+	healthCheckStop := d.healthCheckStop
+
+	go func() {
+		ctx := d.core.activeContext
+		lastTestOk := true
+		lastSeenOk := time.Now()
+		for {
+			select {
+			case <-healthCheckStop:
+				if healthCheck != nil {
+					healthCheck.Stop()
+				}
+				healthCheckStop = nil
+				return
+			case t := <-healthCheck.C:
+				func() {
+					ctx, cancel := context.WithTimeout(ctx, sealHealthTestTimeout)
+					defer cancel()
+
+					testVal := fmt.Sprintf("Heartbeat %d", mathrand.Intn(1000))
+					ciphertext, err := d.Access.Encrypt(ctx, []byte(testVal), nil)
+
+					if err != nil {
+						d.logger.Warn("failed to encrypt seal health test value, seal backend may be unreachable", "error", err)
+						if lastTestOk {
+							healthCheck.Reset(sealHealthTestIntervalUnhealthy)
+						}
+						lastTestOk = false
+						d.core.MetricSink().SetGauge(autoSealUnavailableDuration, 0)
+					} else {
+						plaintext, err := d.Access.Decrypt(ctx, ciphertext, nil)
```

Comment:
Same timeout (shared) across both Encrypt/Decrypt, or should we have a 1m timeout for each?